test/unit/polyfills/test-resize-observer.js

```diff
@@ -0,0 +1,330 @@
+    const parentWin = {
+      ResizeObserver: ResizeObserver1,
+      ResizeObserverEntry: ResizeObserverEntry1,
+    };
+    installForChildWin(parentWin, win);
+    expect(win.ResizeObserver).to.equal(NativeResizeObserver);
+    expect(win.ResizeObserverEntry).to.equal(NativeResizeObserverEntry);
+  });
+});
+
+describes.fakeWin('upgradePolyfill', {}, (env) => {
+  beforeEach(() => {
+    env.sandbox.stub(NativeResizeObserver.prototype, 'observe');
+    env.sandbox.stub(NativeResizeObserver.prototype, 'disconnect');
+  });
+
+  afterEach(() => {
+    resetSubsForTesting();
+  });
+
+  function nextMicroTask() {
+    return Promise.resolve().then(() => Promise.resolve());
+  }
+
+  it('should reset stub before running the installer', () => {
+    const {win} = env;
+    delete win.ResizeObserver;
+    install(win);
+    expect(win.ResizeObserver).to.equal(ResizeObserverStub);
+
+    upgradePolyfill(win, function () {
+      expect(win.ResizeObserver).to.not.be.ok;
+      expect(win.ResizeObserverEntry).to.not.be.ok;
+    });
+  });
+
+  it('should upgrade all stubs', async () => {
+    const {win} = env;
+    delete win.ResizeObserver;
+    install(win);
+
+    const callback = function () {};
+    const element1 = win.document.createElement('div');
+
+    const ro1 = new win.ResizeObserver(callback);
+    const ro2 = new win.ResizeObserver(callback);
+    expect(ro1).to.be.instanceOf(ResizeObserverStub);
+    expect(ro2).to.be.instanceOf(ResizeObserverStub);
+
+    ro1.observe(element1);
+    ro2.observe(element1);
+
+    expect(NativeResizeObserver.prototype.observe).to.not.be.called;
+    upgradePolyfill(win, function () {
+      win.ResizeObserver = NativeResizeObserver;
+      win.ResizeObserverEntry = NativeResizeObserverEntry;
+    });
+
+    await nextMicroTask();
+    expect(NativeResizeObserver.prototype.observe).to.be.calledTwice;
+    expect(NativeResizeObserver.prototype.observe).to.be.calledWith(element1);
+
+    expect(NativeResizeObserver.prototype.disconnect).to.not.be.called;
+    ro1.disconnect();
+    ro2.disconnect();
+    expect(NativeResizeObserver.prototype.disconnect).to.be.calledTwice;
+  });
+
+  it('should auto-upgrade any stub post upgrade', async () => {
+    const {win} = env;
+    delete win.ResizeObserver;
+    install(win);
+    expect(win.ResizeObserver).to.equal(ResizeObserverStub);
+
+    upgradePolyfill(win, function () {
+      win.ResizeObserver = NativeResizeObserver;
+      win.ResizeObserverEntry = NativeResizeObserverEntry;
+    });
+
+    const callback = function () {};
+    const element1 = win.document.createElement('div');
+
+    const ro = new ResizeObserverStub(callback);
+
+    await nextMicroTask();
+    ro.observe(element1);
+    expect(NativeResizeObserver.prototype.observe).to.be.calledOnce;
+    expect(NativeResizeObserver.prototype.observe).to.be.calledWith(element1);
+  });
+
+  it('should run installer even when native is available', () => {
```

Comment:
whats the purpose of this test? this situation shouldn't actually happen right?

Reply:
The same as [this thread ](https://github.com/ampproject/amphtml/pull/31774#discussion_r550332359), I think.